### PR TITLE
Issue 431: return all exceptions as JSON:API error objects

### LINKED ISSUE #326: Document JSON_API_UNIFORM_EXCEPTIONS option
From what I can understand, if this is true, all exceptions raised within the Django project will use JSONAPI formatted responses. That is opposed to the default behavior of only using JSONAPI formatted responses if the view is rendered with JSONAPI.

### LINKED ISSUE #431: spec compliance: exception handling
Split #416 out into exception handling improvements which return useful jsonapi 1.0 [error objects](http://jsonapi.org/format/#error-objects)

### PR DESCRIPTION
**Pull request by n2ygk**
Fixes #431 
Fixes #326

## Description of the Change

Improve exception handling to return JSON:API [error](http://jsonapi.org/format/#errors) objects in all cases.

## Checklist

- [x] PR only contains one change (considered splitting up PR)
- [x] unit-test added
- [x] documentation updated
- [x] changelog entry added to `CHANGELOG.md`
- [x] author name in `AUTHORS`

### CODE REVIEW

**Comment by n2ygk** [comment]
I would appreciate some suggestions on how to cleanly implement test coverage for this.

**Comment by n2ygk** [comment]
(I feel a little guilty about sneaking in coverage to tox.ini as part of this PR....)

**Comment by n2ygk** [comment]
@sliverc Please take a look at my workaround for py27 mock not working the same as py36's. I think I'm doing it correctly. Meanwhile I've wrapped it with xfail as well.

Subject to any suggestions you have, this is ready for review.

File: example/tests/test_views.py [hunk 1]
```diff
@@ -280,3 +289,32 @@ def test_no_content_response(self):
...
+    def get_queryset(self, *args, **kwargs):
+        return self.queryset.filter(foo=1)
+
+
+@pytest.mark.xfail((sys.version_info.major, sys.version_info.minor) == (2, 7),
+                   reason="python2.7 mock raises "
+                   "'unbound method get_queryset() must be called with BadBlogViewSet "
+                   "instance as first argument (got nothing instead)' "
+                   "rather than working properly as does python3.x mock")
+class TestViewExceptions(TestBase):
```
[the remark this answers is not in the document]
Comment (n2ygk) [reply example/tests/test_views.py]: I was frankly confused by the 3-levels of tests/, tests/unit, tests/integration. Any kind of more detail in the contributing doc you could add would be appreciated.

I can move BadBlogViewSet into the example app views. I based the mocking approach on other projects I've seen where the tests are self-contained in the test code.

File: rest_framework_json_api/exceptions.py [hunk 2]
```diff
@@ -14,6 +18,52 @@ def rendered_with_json_api(view):
...
+    Deal with exceptions that DRF doesn't catch and return a JSON:API errors object.
+
+    For model query parameters, attempt to identify the parameter that caused the exception:
+    "Cannot resolve keyword 'xname' into field. Choices are: name, title, ...."
+    Unfortunately there's no "clean" way to identify which field caused the exception other than
+    parsing the error message. If the parse fails, a more generic error is reported.
+
+    Even a 500 error for a jsonapi view should return a jsonapi error object, not an HTML response.
+    """
+    if not rendered_with_json_api(context['view']):
```
[the remark this answers is not in the document]
Comment (n2ygk) [reply rest_framework_json_api/exceptions.py]: Sure, will take a look.

File: rest_framework_json_api/exceptions.py [hunk 3]
```diff
@@ -14,6 +18,52 @@ def rendered_with_json_api(view):
...
+        return None
+    keymatch = re.compile(r"^Cannot resolve keyword '(?P<keyword>\w+)' into field.")
+    if isinstance(exc, (FieldError, FieldDoesNotExist,)):
+        matched = keymatch.match(str(exc))
+        bad_kw = matched.group("keyword") if matched else "?"
+        status_code = 400
+        errors = [
+            {
+                "status": status_code,
+                "code": "field_error",
```
[the remark this answers is not in the document]
Comment (n2ygk) [reply rest_framework_json_api/exceptions.py]: I'd rather err on the side of providing more information to the client (e.g. `field_error` provides more useful information than just a 4xx status code).

File: rest_framework_json_api/exceptions.py [hunk 4]
```diff
@@ -14,6 +18,52 @@ def rendered_with_json_api(view):
...
+    else:
+        status_code = 500
+        errors = [
+            {
+                "status": status_code,
+                "code": "server_error",
+                "title": "Internal Server Error",
+                "detail": str(exc),
+                "source": {
+                    "pointer": context["request"].path,
```
[the remark this answers is not in the document]
Comment (n2ygk) [reply rest_framework_json_api/exceptions.py]: See prior comment; When this is exception is raised as part of a viewset GET it is (maybe?) caused by an incorrect field name.

**Comment by n2ygk** [comment]
FYI - I'll be traveling for the next 2 weeks so will not get a chance to look at this beyond this weekend in case I appear to "go silent". I will be back!

Thanks for all the help.

**Comment by sliverc** [comment]
@n2ygk 
As you have already commented on my inline comments I hope you saw https://github.com/django-json-api/django-rest-framework-json-api/pull/437#pullrequestreview-121309166 which is more important that we discuss it before you continue working on it.

However we can discuss it once you are back. Enjoy your time away!

**Comment by n2ygk** [comment]
@sliverc so it's been a little more than 2 weeks(!) since I had other things backed up when I returned as we are in the midst of developing a new set of microservices based on DJA and django-oauth-toolkit, among other things. I'm going to start by digging in to your recommendation of using `rest_framework.filters` to achieve the functionality I'm trying to achieve in #432 and then circle back to any required improvements in exception handling that leads to.

**Comment by n2ygk** [comment]
I'm closing the PR with the plan to open a new one once I figure out a better way to do this.

**Pull request closed without merging**